File: hs/src/Test/Q6_v1.hs
module Test.Q6_v1 where

import FDB.RustFFI

type FDBInt = CUInt32

data LineItem = LineItem {
  -- l_orderref      :: TableRef Order,
  -- l_partref       :: TableRef Part,
  -- l_suppref       :: TableRef Supplier,
  -- l_linenumber    :: Int,
  l_quantity      :: FDBInt, -- Decimal,
  l_extendedprice :: FDBInt, -- Decimal,
  l_discount      :: FDBInt, -- Decimal,
  -- l_tax           :: Decimal,
  -- l_returnflag    :: Char,
  -- l_linestatus    :: Char,
  l_shipdate      :: FDBInt -- Date,
  -- l_commitdate    :: Date,
  -- l_receiptdate   :: Date,
  -- l_shipinstruct  :: String,
  -- l_shipmode      :: String,
  -- l_comment       :: String
}

-- q6 dbCtx oldDate tgtDiscount maxQty =
q6 dbCtx =
  do
    allItems <- readT dbCtx "lineitem"
    itemsInDateRange <- filterQ inDateRange "Test.Q6_v1.inDateRange" allItems
    itemsWithDiscount <- filterQ discountAroundTarget "Test.Q6_v1.discountAroundTarget" itemsInDateRange
    consideredItems <- filterQ lessQuantity "Test.Q6_v1.lessQuantity" itemsWithDiscount
    foldQ sumExpectedPrice "Test.Q6_v1.sumExpectedPrice" zero "Test.Q6_v1.zeroi" consideredItems
    -- sumOf plannedPrice consideredItems
  -- sumOf f q = foldQ (\x y -> x + f y) "abc" zero "Test.Q6_v1.zeroi" q
  -- sumOf f q = foldQ (\x y -> x + f y) fooFoldZero1 q
  -- sumOf f q = mapAgg sumAgg f q
  -- itemsInDateRange = filterQ inDateRange "Test.Q6_v1.inDateRange" allItems
  -- itemsInDateRange = filterQ inDateRange allItems
  -- consideredItems = filterQ lessQuantity itemsWithDiscount
  -- itemsWithDiscount = filterQ discountAroundTarget itemsInDateRange

-- inDateRange oldDate l =
inDateRange l =
  let
    oldDate = 1
  in
    l_shipdate l <= oldDate + 1
    -- (l_shipdate l >= oldDate) && (l_shipdate l < oldDate + 1)
-- inDateRange oldDate l = (l_shipdate l >= oldDate) && (l_shipdate l < oldDate + 1)
-- inDateRange l = (l_shipdate l >= oldDate) && (l_shipdate l < oldDate `plusInterval` Years 1)

-- discountAroundTarget tgtDiscount l =
discountAroundTarget l =
  let
    tgtDiscount = 2
    discountRangeRadius = 1
    -- discountRangeRadius = 0.01
  in
    l_discount l <= tgtDiscount + discountRangeRadius
    -- (l_discount l >= tgtDiscount - discountRangeRadius)
    -- && (l_discount l < tgtDiscount + discountRangeRadius)

-- lessQuantity maxQty l = l_quantity l < maxQty
lessQuantity l =
  let
    maxQty = 3
  in
    l_quantity l <= maxQty
    -- l_quantity l < maxQty

sumExpectedPrice s l =
  let
    expectedPrice = l_extendedprice l + l_discount l
    -- expectedPrice = l_extendedprice l * l_discount l
  in
    s + expectedPrice
-- (\x y -> x + f y) plannedPrice

zero :: FDBInt
zero = fromIntegral zeroi

zeroi :: Int
zeroi = 0

File: doc/proto1/q6_v1.rs
const L_QUANTITY_IDX = 0;
const L_EXTENDEDPRICE_IDX = 1;
const L_DISCOUNT_IDX = 2;
const L_SHIPDATE_IDX = 3;

fn main () {
  /* foldQ: q6_v1 PREFIX */
  let mut s = 0;
  /* readT: allItems */
  let nrows = count_rows("lineitem_v1");
  for rowid in 0..nrows {
    let fields = read_row("lineitem_v1", rowid);
    /* filterQ: itemsInDateRange */
    let old_date = 19960313;
    let l_shipdate = fields.get(L_SHIPDATE_IDX).unwrap();
    let b1 = l_shipdate <= old_date + 1;
    if b1 {
      /* filterQ: itemsWithDiscount */
      let tgtDiscount =9;
      let discountRangeRadius= 1;
      let l_discount = fields.get(L_DISCOUNT_IDX).unwrap();
      let b2 = l_discount <= tgtDiscount + discountRangeRadius;
      if b2 {
        /* filterQ: consideredItems */
        let maxQty = 32;
        let l_quantity = fields.get(L_QUANTITY_IDX).get();
        let b3 = l_quantity <= maxQty;
        if b3 {
          /* foldQ: q6_v1 MAIN BODY */
          let l_extendedprice = fields.get(L_EXTENDEDPRICE_IDX).get();
          let l_discount =  fields.get(L_DISCOUNT_IDX).get();
          let expectedPrice = l_extendedprice + l_discount;
          s = s + expectedPrice;
        }
      }
    }
  }
  /* foldQ: q6_v1 SUFFIX */
  println!("Result: {}", s);
}
